Core/include/Acts/Seeding/GNN_TrackingFilter.hpp

```diff
@@ -0,0 +1,387 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+// TODO: update to C++17 style
+#include "Acts/Seeding/GNN_DataStorage.hpp"  //includes geo which has trigindetsilayer, may move this to trigbase
+
+#include <algorithm>
+#include <cmath>
+#include <cstring>
+#include <iostream>
+#include <list>
+#include <vector>
+
+template <typename external_spacepoint_t>
+struct TrigFTF_GNN_EdgeState {
+ public:
+  struct Compare {
+    bool operator()(const struct TrigFTF_GNN_EdgeState* s1,
+                    const struct TrigFTF_GNN_EdgeState* s2) {
+      return s1->m_J > s2->m_J;
+    }
+  };
+
+  TrigFTF_GNN_EdgeState(){};
+
+  TrigFTF_GNN_EdgeState(bool f) : m_initialized(f){};
+
+  ~TrigFTF_GNN_EdgeState(){};
+
+  void initialize(Acts::TrigFTF_GNN_Edge<external_spacepoint_t>* pS) {
+    m_initialized = true;
+
+    m_J = 0.0;
+    m_vs.clear();
+
+    // n2->n1
+
+    float dx = pS->m_n1->m_sp_FTF.SP->x() - pS->m_n2->m_sp_FTF.SP->x();
+    float dy = pS->m_n1->m_sp_FTF.SP->y() - pS->m_n2->m_sp_FTF.SP->y();
+    float L = std::sqrt(dx * dx + dy * dy);
+
+    m_s = dy / L;
+    m_c = dx / L;
+
+    // transform for extrapolation and update
+    //  x' =  x*m_c + y*m_s
+    //  y' = -x*m_s + y*m_c
+
+    m_refY = pS->m_n2->m_sp_FTF.SP->r();
+    m_refX =
+        pS->m_n2->m_sp_FTF.SP->x() * m_c + pS->m_n2->m_sp_FTF.SP->y() * m_s;
+
+    // X-state: y, dy/dx, d2y/dx2
+
+    m_X[0] =
+        -pS->m_n2->m_sp_FTF.SP->x() * m_s + pS->m_n2->m_sp_FTF.SP->y() * m_c;
+    m_X[1] = 0.0;
+    m_X[2] = 0.0;
+
+    // Y-state: z, dz/dr
+
+    m_Y[0] = pS->m_n2->m_sp_FTF.SP->z();
+    m_Y[1] = (pS->m_n1->m_sp_FTF.SP->z() - pS->m_n2->m_sp_FTF.SP->z()) /
+             (pS->m_n1->m_sp_FTF.SP->r() - pS->m_n2->m_sp_FTF.SP->r());
+
+    memset(&m_Cx[0][0], 0, sizeof(m_Cx));
+    memset(&m_Cy[0][0], 0, sizeof(m_Cy));
```

Comment:
This is a trial for commenting on already-merged code as a prompt for future improvement.

Please replace `memset` by `std::fill`. `memset` here is undefined behaviour, but works because `0.0f` is represented by 4 `'\0'`s. A bit awkwardly, you need to specify the number of elements as `sizeof(m_Cx)/sizeof(m_Cx[0][0])`.

If `m_Cx` were a `std::array`, then it would know its own size and all would be good.